client/src/actions/annotation.ts

```diff
@@ -0,0 +1,643 @@
+      }
+    } catch (error) {
+      console.error("Failed to rename category on backend:", error);
+      // Same as above - don't revert local state
+    }
+
+    // Reset autosave state AFTER backend completes to prevent autosave from trying
+    // to save to the old category name while the backend file is being renamed
+    dispatch({
+      type: "writable obs annotations - save complete",
+      lastSavedAnnoMatrix: obsCrossfilter.annoMatrix,
+    });
+  };
+
+export const annotationDeleteCategoryAction =
+  (categoryName: string): ThunkResult =>
+  async (dispatch: AppDispatch, getState: GetState) => {
+    const state = getState();
+    const {
+      annoMatrix: prevAnnoMatrix,
+      obsCrossfilter: prevObsCrossfilter,
+      config,
+    } = state;
+    if (!prevAnnoMatrix || !prevObsCrossfilter) return;
+    if (!isUserAnnotation(prevAnnoMatrix, categoryName)) {
+      throw new Error("not a user annotation");
+    }
+
+    // Update in-memory state IMMEDIATELY for optimistic UI update
+    const obsCrossfilter = prevObsCrossfilter.dropObsColumn(categoryName);
+    dispatch({
+      type: "annotation: delete category",
+      annoMatrix: obsCrossfilter.annoMatrix,
+      obsCrossfilter,
+      metadataField: categoryName,
+    });
+
+    // Only make backend request if annotations are enabled (not in no-auth mode)
+    const writableCategoriesEnabled = !!config?.parameters?.annotations;
+    if (!writableCategoriesEnabled) {
+      // In no-auth mode, immediately reset autosave state since there's no backend to wait for
+      dispatch({
+        type: "writable obs annotations - save complete",
+        lastSavedAnnoMatrix: obsCrossfilter.annoMatrix,
+      });
+      return;
+    }
+
+    // Persist to backend and WAIT for completion
+    try {
+      const response = await fetch(
+        `${globals.API?.prefix ?? ""}${
+          globals.API?.version ?? ""
+        }annotations/obs?category_name=${encodeURIComponent(categoryName)}`,
+        {
+          method: "DELETE",
+          credentials: "include",
+        }
+      );
+
+      if (!response.ok) {
+        console.error(
+          `Failed to delete category on backend: ${response.status} - ${response.statusText}`
+        );
+        // Note: We don't revert the local state here since the user expects the UI to reflect their action
+        // The backend deletion failure will be logged but won't block the UI
+      }
+    } catch (error) {
+      console.error("Failed to delete category on backend:", error);
+      // Same as above - don't revert local state
+    }
+
+    // Reset autosave state AFTER backend completes to prevent autosave from trying
+    // to save the deleted category while the backend file is being deleted
+    dispatch({
+      type: "writable obs annotations - save complete",
+      lastSavedAnnoMatrix: obsCrossfilter.annoMatrix,
+    });
+  };
+
+export const annotationCreateLabelInCategory =
```

Comment:
Label creation now optionally assigns the current selection; nice that we reuse `addObsAnnoCategory`/`setObsColumnValues` so the undo stack stays intact.